src/App.js

```diff
@@ -21,45 +21,116 @@ const generateSquares = () => {
       currentId += 1;
     }
   }
-
   return squares;
 };
 
 const App = () => {
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
-
+  const [player, changePlayer] = useState(PLAYER_1);
+  let [winner, checkWinner] = useState('');
   // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let row = 0;
+    let column = 0;
+    // check for row match
+    while (row < 3) {
+      if (
+        squares[row][0].value === squares[row][1].value &&
+        squares[row][2].value === squares[row][1].value &&
+        squares[row][0].value != ''
+      ) {
+        return squares[row][0].value;
+      } else row += 1;
+    }
+    // column check
+    while (column < 3) {
+      if (
+        squares[0][column].value === squares[1][column].value &&
+        squares[2][column].value === squares[1][column].value &&
+        squares[0][column].value != ''
+      ) {
+        return squares[0][column].value;
+      } else column += 1;
+    }
+    // diagonal check
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[0][0] != ''
+    ) {
+      return squares[0][0].value;
+    } else if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[0][2] != ''
+    ) {
+      return squares[0][2].value;
+    } else {
+      for (let arr of squares) {
+        for (let obj of arr) {
+          if (obj.value == '') {
+            return '';
+          }
+          continue;
+        }
+      }
+      return 'Tie';
+    }
   };
 
+  const onClickCallback = (id) => {
+    let square = squares.map((oneSquareArray) => {
+      for (const insideSquare of oneSquareArray) {
+        if (insideSquare.id === id) {
+          if (insideSquare.value) {
+            continue;
+          } else {
+            insideSquare.value = player;
+          }
+        }
+      }
+      return oneSquareArray;
+    });
```

Comment:
Great use of `map` to iterate and return a new array!

We can combine the conditionals in the inner loop using compound expressions and the not operator `!` like so:

```suggestion
    let square = squares.map((oneSquareArray) => {
      for (const insideSquare of oneSquareArray) {
        if (insideSquare.id === id && !insideSqaure.value) {
            insideSquare.value = player;
        }
      }
      return oneSquareArray;
    });
```